### DISCUSSION TITLE
Javascript not working?

### DISCUSSION THREAD

**Discussion by ghost** [community/community#21474] [Other Feature Feedback, Questions, & Ideas]
<p>Javascript not working! I tried the following:</p>
<ul>
<li>Writing <code>&lt;!DOCTYPE html&gt;&lt;html&gt;&lt;script&gt;...&lt;/script&gt;&lt;/html&gt;</code> in a <code>.html</code> file</li>
<li>Writing <code>&lt;!DOCTYPE html&gt;&lt;script&gt;...&lt;/script&gt;</code> in a <code>.html</code> file</li>
<li>Writing <code>...</code> in a <code>.js</code> file<br>
I checked the languages, but all it says is:</li>
</ul>
<pre><code class="lang-auto">100% html</code></pre>
<p>Who can help me?</p>

**Comment by thinkverse** [comment]
<p>Hi. 👋</p>
<p>There are several ways to include JavaScript in an HTML file, the most used are internal or external, both internal and external are usually placed either in the <code>&lt;head&gt;</code> or the last thing before closing the <code>&lt;/body&gt;</code>.</p>
<pre><code class="lang-auto">&lt;!DOCTYPE html&gt;
&lt;html&gt;
   &lt;head&gt;
      &lt;title&gt;Your title&lt;/title&gt;
      &lt;script type="text/javascript"&gt;
            let variable = 'I am internal'
      &lt;/script&gt;
      &lt;script src="./js/external-01.js" type="text/javascript"&gt;
   &lt;/head&gt;
   &lt;body&gt;
      ...content

      &lt;script src="./js/external-02.js" type="text/javascript"&gt;
      &lt;script type="text/javascript"&gt;
            let variable2 = 'I am also internal'
      &lt;/script&gt;
   &lt;/body&gt;
&lt;/html&gt;
</code></pre>
<p>Hope that helps your journey into the land of code. :+1:</p>

**Comment by ghost** [comment]
<p>My answer: <span class="bbcode-b">This won’t work.</span><br>
Proof: <span class="bbcode-b">It says 100% HTML.</span><br>
Evidence: <a href="https://github.com/BosenChang/bosenchang.github.io" rel="nofollow noopener">This</a> :angry:</p>

**Comment by airtower-luna** [comment]
<p>The language guess on your repository page says precisely nothing about whether the script will work. :slightly_smiling_face:</p>
<p>To elaborate a bit more: Github uses Linguist to guess the language of your files. Those guesses are strictly <em>per file</em>, an HTML file that includes a few scripts will show up as HTML. If you keep your scripts in separate files and include those from your HTML file, the separate script files should show up as Javascript.<br>
</p><aside class="onebox allowlistedgeneric">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">
      <a href="https://github.com/github/linguist" target="_blank" rel="noopener">GitHub</a>
  </header>
  <article class="onebox-body">
    <img src="https://user-images.githubusercontent.com/5814919/181064910-ad1b444d-5420-4cfc-9039-d2a804264c5b.png" class="thumbnail onebox-avatar" width="400" height="400">

<h3><a href="https://github.com/github/linguist" target="_blank" rel="noopener">github/linguist</a></h3>

<p>Language Savant. If your repository's language is being reported incorrectly, send us a pull request! - github/linguist</p>


  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by ghost** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="4" data-topic="131042">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>The language guess on your repository page says precisely nothing about whether the script will work. :slightly_smiling_face:</p>
<p>To elaborate a bit more: Github uses Linguist to guess the language of your files. Those guesses are strictly <em>per file</em> , an HTML file that includes a few scripts will show up as HTML. If you keep your scripts in separate files and include those from your HTML file, the separate script files should show up as Javascript.</p>
</blockquote>
</aside>
<p>Could you explain it clearer?</p>

**Comment by thinkverse** [comment]
<p>What you have in your .html files aren’t 100% valid HTML structure, it’s missing the <code>&lt;head&gt;</code>  and <code>&lt;body&gt;</code> tags, GitHub doesn’t check for valid structure, it guesses.</p>
<p>What I posted above is a valid HTML structure, the content that you display on your page should be inside the <code>&lt;body&gt;</code> tag, that’s the <em>body</em> of your page so-to-speak, inside the <code>&lt;head&gt;</code> is your metadata, like your <em>page title</em> -&gt; <code>&lt;title&gt;My HTML page&lt;/title&gt;</code>, links to your stylesheets <code>&lt;link rel="stylesheet" href="css/my-css-file.css"&gt;</code> etc.</p>
<p>The <code>&lt;html&gt;</code> tags wrap around those tags, making it a valid tree structure, the <code>&lt;DOCTYPE html&gt;</code> is there for browsers and IDE’s to interpret what type of file this is and what standard it should use when rendering the page, <code>&lt;DOCTYPE html&gt;</code> is the document type for HTML5, aka <a href="https://html.spec.whatwg.org/" rel="nofollow noopener">HTML living standard</a>.</p>
<p>Hope that explains it. :slightly_smiling_face:</p>

**Comment by ghost** [comment]
<p>No, not really.<br>
Here is the code:</p>
<pre><code class="lang-auto">&lt;!DOCTYPE html&gt;
&lt;html&gt;
&lt;head&gt;
&lt;title&gt;
Password Protected
&lt;/title&gt;
&lt;script type="text/javascript"&gt;
alert("Who are you?");
alert("Sign in information:");
var password = prompt("Password?");
if (password) == ("HIDDEN123") {
alert("Access granted");
}else{
history.back();
}
&lt;/script&gt;
&lt;/head&gt;
&lt;body&gt;
&lt;/body&gt;
&lt;/html&gt;
</code></pre>
<p>That code is the one I used.</p>

**Comment by thinkverse** [comment, marked as answer]
<p>If you’re using Chrome or Firefox you should have access to a developer console, which can be accessed in many ways, but usually, you can right-click a page and choose the <code>inspect</code> option, or the equivalent in your language, if you get that open and click the console, your see the following error.</p>
<pre><code class="lang-bash">Uncaught SyntaxError: Unexpected token '==' password_protected.html:11 
</code></pre>
<p>Which means there’s a syntax error in your file on line 11, in this case, that’s this line</p>
<pre><code class="lang-auto">if (password) == ("HIDDEN123") {
</code></pre>
<p>A syntax error will error out your compiler meaning whatever’s in there won’t run, change the line to <code>if (password == "HIDDEN123") {</code> and it will work. :slightly_smiling_face:</p>
